Refactor the PSF Photometry API

barentsen opened the pull request:
@christinahedges and @geert pair-coded a non-functional doodle of our dream API for PSF photometry.  The idea is to re-factor @mirca's existing PSF-fitting code to make it slightly easier to use.

The core of the idea is that we should have a `tpf.psf_photometry(catalog="kic")` function that returns a collection of `LightCurve` objects.

This is very much work in progress and just a doodle!

mirca commented:
This is KEY for an intuitive/user-friendly PSF photometry API, thanks for putting y'all's thoughts in the open ❤️

gully commented:
The doodle looks good to me.  In my experience with inference frameworks, friction arises when users wish to extend the model with more tunable knobs.  I don't catch sight of anything that limits extensibility.  

Presumably most users will want photometry, but some might also find value in the fitted (x,y) centroids, or future added fit parameters, if any.

christinahedges commented:
I think we need some ability to "freeze" parameters to stop them being fit. This will then overcome the issue of fitting a single target for RA/Dec and Motion (which is degenerate), because we'll freeze the RA/Dec.

mirca commented:
@christinahedges I think a nice and simple way to freeze parameters would be to create another instance of your model. Say your model parameters are ``x, y, flux`` and you call it using ``psf_model(x, y, flux)``, then to freeze, say, the centroids, you could do ``psf_model_flux = lambda flux: psf_model(x=1., y=2., flux=flux)``

christinahedges commented:
@mirca that sounds good, do we think users want the ability to freeze and thaw parameters as they go? Something like...

`PSF = psfmodel(priors=priors)`
`PSF.freeze([RA, Dec])`
`PSF.fit()`

Perhaps this isn't super necessary. Perhaps they can specify which are frozen like this:

`PSF = psfmodel(priors=priors, freeze=[RA, Dec])`
`PSF.fit()`

mirca commented:
@christinahedges Yeah, I think having a keyword option to freeze the parameters is very convenient.

barentsen commented:
**Update: we have a working prototype!**

The snippet below shows an example of what the new API in this PR currently looks like.  It demonstrates the main goal of this PR, which is to introduce high-level classes such as `StarPrior` and `BackgroundPrior` which make it easier to setup a model.  There can also be a `MotionPrior` and `FocusPrior` once the simple version produces reasonable lightcurves.

@mirca @amcody @gully please post your thoughts!
(@mirca we lost gradients for now 😢)

```Python
def psf_photometry_demo():

    # We will attempt to make a lightcurve for Tabby's star
    tpf = KeplerTargetPixelFile.from_archive(8462852, quarter=16, quality_bitmask='hardest')

    # First, set up a simple scene model with one star and no motion or focus changes
    star_prior = StarPrior(flux=np.nansum(tpf.flux[0]), col=tpf.column, row=tpf.row,
                           err_flux=1e5, err_col=2, err_row=2,
                           targetid=tpf.keplerid)
    background_prior = BackgroundPrior(flux=np.nanpercentile(tpf.flux[0], 10), err_flux=100.0)
    scenemodel = SimpleSceneModel(star_priors=[star_prior],
                                  background_prior=background_prior,
                                  prfmodel=tpf.get_prf_model())

    # Now make the lightcurve by fitting each cadence
    time = tpf.time[1400:1700]
    flux = []
    for idx in tqdm(range(len(time))):
        result = scenemodel.fit(tpf.flux[idx])
        flux.append(result.stars[0].flux)
    return LightCurve(time, flux)
```

mirca commented:
@barentsen This looks exciting to me!!

christinahedges commented:
The refactored code currently doesn't work properly. If you compare to the original PRF photometry on tabby's star it gives completely the wrong answer.

![demo](https://user-images.githubusercontent.com/14965634/41561027-a788a0c0-72fd-11e8-8182-54a73c7ae33f.png)

christinahedges commented:
Alright I fixed it. 

![demo](https://user-images.githubusercontent.com/14965634/41568394-5118c9f0-731a-11e8-8e68-b37fc69b0cfc.png)

@barentsen there was an error in the way you wrong your test, 

This 
```
time = tpf.time[1400:1700]  # this cadence range include the deepest dip
for idx in tqdm(range(len(time))):
```
should be this

```
time = tpf.time[1400:1700]  # this cadence range include the deepest dip
for idx in tqdm(np.arange(1400, 1700)):
```

This took me ages to find it's not even funny. I did fix a different bug in the mean time with the _predict and from_array functions. All fixed now!

mirca commented:
@christinahedges zoodles points as well; that was a sick finding, Chedges!

barentsen commented:
Update: I added the ability to fit the telescope focus and motion parameters.  Fitting these is optional and can easily be switched on/off via the `fix_motion` and `fix_focus` flags.   The API now looks like this:

```Python
%matplotlib inline 
import numpy as np
from lightkurve import KeplerTargetPixelFile, LightCurve
from lightkurve.prf_photometry import StarPrior, FocusPrior, MotionPrior, SceneModel, PRFPhotometry
from oktopus import GaussianPrior, UniformPrior

tpf = KeplerTargetPixelFile("https://archive.stsci.edu/missions/kepler/target_pixel_files/0084/008462852/"
                            "kplr008462852-2013098041711_lpd-targ.fits.gz", quality_mask='hardest')

# First, compute a few values from our TPF which will inform the priors
bgflux = np.nanpercentile(tpf.flux[0], 10)
maxflux = np.nansum(tpf.flux, axis=(1, 2)).max()
col, row = np.nanmedian(tpf.centroids(), axis=1)

# Set up the model
model = SceneModel(star_priors=[StarPrior(col=GaussianPrior(mean=col, var=2**2),
                                          row=GaussianPrior(mean=row, var=2**2),
                                          flux=UniformPrior(lb=0, ub=maxflux),
                                          targetid=tpf.keplerid)],
                   background_prior=GaussianPrior(mean=bgflux, var=bgflux),
                   focus_prior=FocusPrior(scale_col=GaussianPrior(mean=1, var=0.0001),
                                          scale_row=GaussianPrior(mean=1, var=0.0001),
                                          rotation_angle=UniformPrior(lb=-3.1415, ub=3.1415)),
                   motion_prior=MotionPrior(shift_col=GaussianPrior(mean=0., var=0.01),
                                            shift_row=GaussianPrior(mean=0., var=0.01)),
                   prfmodel=tpf.get_prf_model(),
                   fix_background=False,
                   fix_focus=False,
                   fix_motion=False)

pp = PRFPhotometry(model)
pp.run(tpf.flux[1650:1850], pos_corr1=tpf.pos_corr1[1650:1850], pos_corr2=tpf.pos_corr2[1650:1850])
pp.plot_results()
```

Note: fitting the extra parameters and specifying their priors is entirely optional.  If you wish to rely on the defaults, you can make a much simpler model like this:
```Python
model = SceneModel(star_priors=[StarPrior(col=GaussianPrior(mean=col, var=2**2),
                                          row=GaussianPrior(mean=row, var=2**2),
                                          flux=UniformPrior(lb=0, ub=maxflux),
                                          targetid=tpf.keplerid)],
                   prfmodel=tpf.get_prf_model())
```


Also note: the `pos_corr` data is passed to the PRFPhotometry class which takes care of updating the `MotionPrior` at each cadence using the pos_corr measurements.  (Though I'm not entirely sure yet how much of this logic belongs inside a `PRFPhotometry` class and how much can go inside a `KeplerTargetPixelFile.psf_photometry()` method).

amcody commented:
Nice, I like it! Gonna take a closer look and see if we can run it on some K2 sources..

barentsen commented:
Feedback from pair-chatting with @christinahedges:
* classes like MotionPrior, FocusPrior, etc... should accept fixed values for their parameters (which, behind the scenes, may or may not be represented as a delta distribution);
* consider renaming `fix_focus` into `fit_focus` (etc);
* create a `BackgroundPrior` to avoid confusion;
* we need to improve the oktopus reprs;
* we need unit tests;

christinahedges commented:
Does this version give back uncertainties? We have PR #11 open which is supposed to add errors to PSF photometry.

barentsen commented:
It does not return uncertainties yet!  Let's see if we can incorporate #11 into this PR.

This PR is otherwise ready for review, apart from the unit tests needing to become more comprehensive.  It still lacks a `TargetPixelFile.prf_photometry()` function, but we may want to do that in a separate PR once the Collection and StarPrior-generating classes have been meged.

barentsen commented:
@mirca I am attempting to add support for uncertainties, but I am getting confused as to how to implement `SceneModel.gradient()`.  Should it return deltaFlux / deltaParameters like `KeplerPRF.gradient()` does, or should it return deltaProbability / deltaParameters?

My current (poor, incorrect) implementation lives at https://github.com/barentsen/lightkurve/blob/eee2eaa8d0bd5738a1a42bfd5ebf14a5355fc201/lightkurve/prf/scenemodels.py#L457

mirca commented:
It should be the gradient of SceneModel._predict with respect to params. Will
share more thoughts later!

On Thu, Jun 28, 2018 at 5:50 PM, Geert Barentsen <notifications@github.com>
wrote:

> @mirca <https://github.com/mirca> I am attempting to add support for
> uncertainties, but I am getting confused as to how to implement
> SceneModel.gradient(). Should it return deltaFlux / deltaParameters like
> KeplerPRF.gradient() does, or should it return deltaProbability /
> deltaParameters?
>
> My current (poor, incorrect) implementation lives at
> https://github.com/barentsen/lightkurve/blob/
> eee2eaa8d0bd5738a1a42bfd5ebf14a5355fc201/lightkurve/prf/
> scenemodels.py#L457
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/KeplerGO/lightkurve/pull/157#issuecomment-401169053>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMeKO3SB8SxZyCxcw4oF1Lg-MlNTpasqks5uBUGOgaJpZM4UjeTk>
> .
>



-- 
webpage: http://mirca.github.io
github: @mirca <http://www.github.com/mirca>
twitter: @mircaze <http://www.twitter.com/mircaze>

barentsen commented:
The following example gives me a `LinAlgError: Singular matrix` error.  It does work fine with `fit_focus=False`.  This makes sense because `KeplerPRF.gradient()` does not implement the gradient with respect to the focus parameters yet, and therefore there is a hack in `SceneModel.gradient()` to pretend these gradients are all zeroes.  This is probably what leads to the singular matrix...

@mirca Do you think we can add the derivatives with respect to `scale_col`, `scale_row`, and `rotation_angle` to `KeplerPRF.gradient()`?

```Python
from lightkurve import GaussianPrior, UniformPrior
from lightkurve import KeplerPRF
from lightkurve import StarPrior, BackgroundPrior, SceneModel

shape = (12,12)
starflux, bgflux = 1000., 10.
col, row = 60, 70
prf = KeplerPRF(channel=42, shape=(10, 12), column=col, row=row)
star_prior = StarPrior(col=GaussianPrior(col + 6, 0.01),
                       row=GaussianPrior(row + 6, 0.01),
                       flux=UniformPrior(lb=0.5*starflux, ub=1.5*starflux))
background_prior = BackgroundPrior(flux=UniformPrior(lb=0, ub=100))
scene = SceneModel(star_priors=[star_prior],
                   background_prior=background_prior,
                   prfmodel=prf,
                   fit_background=True,
                   fit_focus=True)
# Generate and fit fake data
fake_data = bgflux + prf(starflux, col + 6, row + 6,
                         scale_col=1., scale_row=1.,
                         rotation_angle=0)
results = scene.fit(fake_data, tol=1e-10, options={'maxiter': 100})
```
Error message:
```Python
---------------------------------------------------------------------------
LinAlgError                               Traceback (most recent call last)
<ipython-input-1-a3c8403f8af5> in <module>()
     20                          scale_col=1., scale_row=1.,
     21                          rotation_angle=0)
---> 22 results = scene.fit(fake_data, tol=1e-10, options={'maxiter': 100})

~/dev/lightkurve/lightkurve/prf/scenemodels.py in fit(self, data, loss_function, method, **kwargs)
    527         fit = loss.fit(x0=self.get_initial_guesses().to_array(), method=method, **kwargs)
    528         result = self.params.from_array(fit.x)
--> 529         result.uncertainties = loss.loglikelihood.uncertainties(fit.x)
    530         result.predicted_image = self._predict(fit.x)
    531         result.residual_image = data - result.predicted_image

~/bin/anaconda/lib/python3.6/site-packages/oktopus/likelihood.py in uncertainties(self, params)
     52         Information Matrix
     53         """
---> 54         inv_fisher = np.linalg.inv(self.fisher_information_matrix(params))
     55         return np.sqrt(np.diag(inv_fisher))
     56 

~/bin/anaconda/lib/python3.6/site-packages/autograd/tracer.py in f_wrapped(*args, **kwargs)
     46             return new_box(ans, trace, node)
     47         else:
---> 48             return f_raw(*args, **kwargs)
     49     f_wrapped.fun = f_raw
     50     f_wrapped._is_autograd_primitive = True

~/bin/anaconda/lib/python3.6/site-packages/numpy/linalg/linalg.py in inv(a)
    526     signature = 'D->D' if isComplexType(t) else 'd->d'
    527     extobj = get_linalg_error_extobj(_raise_linalgerror_singular)
--> 528     ainv = _umath_linalg.inv(a, signature=signature, extobj=extobj)
    529     return wrap(ainv.astype(result_t, copy=False))
    530 

~/bin/anaconda/lib/python3.6/site-packages/numpy/linalg/linalg.py in _raise_linalgerror_singular(err, flag)
     87 
     88 def _raise_linalgerror_singular(err, flag):
---> 89     raise LinAlgError("Singular matrix")
     90 
     91 def _raise_linalgerror_nonposdef(err, flag):

LinAlgError: Singular matrix
```

mirca commented:
> and therefore there is a hack in SceneModel.gradient() to pretend these gradients are all zeroes. This is probably what leads to the singular matrix.

Yeah, that will make the determinant of the matrix become zero, so we can't find its inverse.

> @mirca Do you think we can add the derivatives with respect to scale_col, scale_row, and rotation_angle to KeplerPRF.gradient()?

I can happily take a look at that this weekend!

mirca commented:
Oh, and we can't use ``autograd`` to get those gradients because the ``scipy`` interpolation we use is not supported by ``autograd``.

mirca commented on lightkurve/lightcurve.py:
```diff
@@ -47,16 +47,24 @@ class LightCurve(object):
+        if time is None:
+            self.time = np.arange(len(flux))
+        else:
+            self.time = np.asarray(time)
```
If we want to be rigorous here, then let's also check that ``len(flux) == len(time)``?

barentsen replied:
Ok!

barentsen replied:
Oh, `_validate_array` already takes care of checking this,

mirca commented on lightkurve/prf/scenemodels.py:
```diff
@@ -0,0 +1,670 @@
+
+class FixedValuePrior(Prior):
+    """An improper prior with a likelihood probability of 1 at a single fixed
+    value and -inf elsewhere. This is similar to a Dirac Delta function,
```
Since at the top we mention ``probability``, I would replace ``-inf`` with ``0``. Or you can put ``log probability`` at the top and change ``likelihood`` to ``log``.

mirca commented on lightkurve/prf/scenemodels.py:
```diff
@@ -0,0 +1,670 @@
+    def gradient(self, params):
+        if self.value == params:
+            return 0.0
+        return np.inf
```
I'm not sure about that

mirca replied:
The problem is that the gradient of ``UniformPrior`` is undefined at the boundaries. For ``FixedValuePrior`` we don't even have boundaries. My point is that if a parameter is fixed we shouldn't try to compute ``d model / d param``

barentsen replied:
Ok!  I'm going to skip uncertainties for now and raise a `NotImplementedError` for `FixedValuePrior.gradient()`.

mirca commented on lightkurve/prf/scenemodels.py:
```diff
@@ -0,0 +1,670 @@
+        for star in params.stars:
+            star_images.append(self.prfmodel(flux=star.flux,
+                                             center_col=star.col + params.motion.shift_col,
+                                             center_row=star.row + params.motion.shift_row,
```
Is ``star.col + params.motion.shift_col`` necessary? We want the center of the star relative to the TPF, I don't see why we need ``a + b`` to get that. Is this to plug in ``poscorr``? But in the case we want to fit both, aren't these gonna be degenerate?

mirca replied:
Oh, I see here, ``params.motion.shift_col`` is shared among all stars. That looks sweet!! =D

mirca replied:
Still, I don't quite understand this for the case of a single star.

barentsen replied:
Yes, `shift_col` and `shift_row` model the spacecraft motion.  The idea is that we always use very narrow priors on the star positions (because we know star positions very well), and absorb spacecraft motion in `shift_col` and `shift_row`.  This allows us to prevent stars from moving relative to eachother in different cadences, which we know can't happen.

In the case of a single star the user can choose to keep either the motion or the star position fixed.  It is very easy to do this in the new API by setting `fit_motion=False` in the `SceneModel`.

mirca replied:
Makes a lot of sense =)

mirca commented on lightkurve/prf/scenemodels.py:
```diff
@@ -0,0 +1,670 @@
+        grad.append([np.ones(self.prfmodel.shape)])
+        # We assume the gradient of other parameters is zero
+        for i in range(len([*params_array]) - 3 * len(params.stars) - 1):
+            grad.append([np.zeros(self.prfmodel.shape)])
```
We don't want a line/column of zeros in our jacobian matrix xD. Other than that, it looks good, I'm working on adding the gradients for the rest of the parameters.

barentsen replied:
I have to put some values because the size of the gradient needs to match the size of the free parameters.  Shall I put it to one for now?

mirca replied:
Since for now we have gradients available only for ``flux``, ``col``, ``row``, ``shift_row``, ``shift_col``, and ``bkg``, I would probably assume that the other parameters are fixed

mirca commented on lightkurve/prf/scenemodels.py:
```diff
@@ -0,0 +1,670 @@
+        self.model = model
+        self.results = []
+
+    def run(self, tpf_flux, pos_corr1=None, pos_corr2=None):
```
I think it's worth adding a parameter ``cadences`` here in case users want to fit only a subset of cadences?

barentsen replied:
Done!

mirca commented on lightkurve/prf/tests/test_scenemodels.py:
```diff
@@ -0,0 +1,159 @@
+    assert np.isclose(results.background.flux, bgflux)
+    assert np.isclose(results.focus.scale_col, scale_col)
+    assert np.isclose(results.focus.scale_row, scale_row)
+    assert np.isclose(results.focus.rotation_angle, rotation_angle)
```
Sweet unit tests!

mirca commented:
Other than removing the array of zeros in the Jacobian matrix, I think this is ready to be merged! 🎉🎉🎉🎉

barentsen commented on lightkurve/prf/scenemodels.py:
```diff
@@ -0,0 +1,670 @@
+
+
+class FixedValuePrior(Prior):
+    """An improper prior with a likelihood probability of 1 at a single fixed
```
Ok!  We may want to move a class like this to oktopus eventually -- I kept it in lightkurve for now just because it is easier to develop.

AndyTza commented:
<img width="881" alt="screen shot 2018-07-10 at 11 04 21 am" src="https://user-images.githubusercontent.com/22720676/42528677-169610da-8431-11e8-8d2f-db87b56e01bf.png">

@christinahedges and I came across this issue yesterday. It turns out that `prfmodel=tpf.get_prf_model() `couldn't sucessfully find the tpf model due to keyword argument.

We temporarily fixed it by mannually calling KeplerPRF to generate the PRF model.
`prfmodel=KeplerPRF(tpf.channel, tpf.flux.shape[1:], tpf.column, tpf.row)`

Could we warn the users if the PRF model is used the wrong way? 
Also, in `tpf.get_prf_model()` we could maybe add a keyword argument `simple=True` to choose method by default.

barentsen commented:
This is probably a trivial error.  I have to run but will try to fix later.  Feel free to push a commit or open a PR against this branch if you find the fix before I'm back/

barentsen commented:
@AndyTza I am unable to re-produce the error you posted above.  Can you double-check that you are using a clean version of the `refactor-psf-photometry` branch?  If you still have a `prf_photometry.py` file in your tree (which should no longer be there), then try `git reset --hard` to remove all outdated files.  (Warning: this will also remove any other files you may be working on but haven't committed ... if you are worried about that, then please check out the branch in some new directory.)

If the problem keeps happening, please copy/paste a full failing example below.

PS: the API now works as follows (note the import from `lightkurve.prf` instead of `lightkurve.prf_photometry` in this updated example):
```
%matplotlib inline
import numpy as np
from lightkurve import KeplerTargetPixelFile, LightCurve
from lightkurve.prf import StarPrior, BackgroundPrior, FocusPrior, MotionPrior, SceneModel, PRFPhotometry
from oktopus import GaussianPrior, UniformPrior

tpf = KeplerTargetPixelFile("https://archive.stsci.edu/missions/kepler/target_pixel_files/0084/008462852/"
                            "kplr008462852-2013098041711_lpd-targ.fits.gz", quality_mask='hardest')

# First, compute a few values from our TPF which will inform the priors
bgflux = np.nanpercentile(tpf.flux[0], 10)
maxflux = np.nansum(tpf.flux, axis=(1, 2)).max()
col, row = np.nanmedian(tpf.centroids(), axis=1)

# Set up the model
model = SceneModel(star_priors=[StarPrior(col=GaussianPrior(mean=col, var=2**2),
                                          row=GaussianPrior(mean=row, var=2**2),
                                          flux=UniformPrior(lb=0, ub=maxflux),
                                          targetid=tpf.keplerid)],
                   background_prior=BackgroundPrior(flux=GaussianPrior(mean=bgflux, var=bgflux)),
                   focus_prior=FocusPrior(scale_col=GaussianPrior(mean=1, var=0.0001),
                                          scale_row=GaussianPrior(mean=1, var=0.0001),
                                          rotation_angle=UniformPrior(lb=-3.1415, ub=3.1415)),
                   motion_prior=MotionPrior(shift_col=GaussianPrior(mean=0., var=0.01),
                                            shift_row=GaussianPrior(mean=0., var=0.01)),
                   prfmodel=tpf.get_prf_model(),
                   fit_background=True,
                   fit_focus=False,
                   fit_motion=False)

pp = PRFPhotometry(model)
pp.run(tpf.flux, pos_corr1=tpf.pos_corr1, pos_corr2=tpf.pos_corr2, cadences=range(1650, 1850))
pp.plot_results()
print('The star flux in the first cadence is {}'.format(pp.results[0].stars[0].flux))
```

barentsen commented:
I renamed `SceneModel` into `TPFModel`, so the example above is outdated again.  (You can see the updated example at the top of `lightkurve/prf/tpfmodel.py`.

I implemented a convenience function which uses the API with some default assumptions as follows:
```Python
from lightkurve import KeplerTargetPixelFile
tpf = KeplerTargetPixelFile("kplr008462852-2013098041711_lpd-targ.fits.gz", quality_mask='hardest')
pp = tpf.prf_photometry(cadences=range(1650, 1800))
pp.plot_results()
```

You can also simply do:
```Python
lc = tpf.to_lightcurve(method='prf')
```

...but please don't assume it will work great out of the box yet, we'll have to tune the default assumption on the priors.

Bonus: `PRFPhotometry.run()` now fits cadences in parallel using `multiprocessing`.

This is getting really close to being merged.  Please review!

mirca commented on lightkurve/targetpixelfile.py:
```diff
@@ -826,6 +861,64 @@ def get_bkg_lightcurve(self, aperture_mask=None):
+        ----------
+        cadences : list of int
+            Cadences to fit.  This can be useful while debugging a subset of
+            lightcurve cadences.
```
I was thinking whether ``fit_parallel`` would fit better than ``use_multiprocessing``.
In any case, this is more of a reminder to add docstring for this kw argument.

mirca replied:
Let's explain what happens when ``cadences is None`` here, as well?

barentsen replied:
Yes, good idea!

mirca commented:
I can confirm that the example in ``tpfmodel.py`` runs smoothly. (I tried different combinations of ``fit_focus``, ``fit_motion``, ``fit_background``)

barentsen commented:
Thanks @mirca !!  I expect to merge this later today after a few final improvements.

There is some work to do. In particular, we need tutorials, we need to add uncertainties, and we need to think about how to include a data-driven PSF approach, e.g. using tensorflow.  But we can develop those technologies in the background and in separate packages.  In meanwhile, @amcody & friends can start experimenting with this API!

mirca commented on lightkurve/prf/tpfmodel.py:
```diff
@@ -406,7 +406,7 @@ def __init__(self, star_priors=[],
                  motion_prior=MotionPrior(),
                  prfmodel=KeplerPRF(1, shape=(10, 10), column=0, row=0),
-                 fit_background=True, fit_focus=False, fit_motion=False):
+                 fit_background=False, fit_focus=False, fit_motion=False):
```
why not fit the background by default? I think it's a common thing to add the background back (to avoid negative flux values) and fit it (Libralato et. al. does that).

barentsen commented:
@mirca YOU ARE MY HERO!!  The background-fitting performed really poorly while I was writing a tutorial yesterday, so I turned it off by default for now.  However: I was fitting `tpf.flux` instead of `tpf.flux + tpf.flux_bkg`.  Thanks for the suggestion to use the latter, that does work way better!

Turning background fitting back on by default when I push the next commit.  Thanks++++. :+1: :+1: :+1:

barentsen commented:
I added a first PRF photometry tutorial (`2.10-how-to-use-prf-photometry.ipynb`), and convinced myself that the API is good enough to be merged as an experimental feature ahead of the start of today's K2 Supernova Workshop!   When PSF photometry is used, a warning message is sent to `logging` regarding the experimental nature.

I am going to merge this now and start separate PRs on future improvements, most notable the addition of support for uncertainties (cf. https://github.com/barentsen/lightkurve/pull/2 )

mirca commented:
@barentsen the AppVeyor build seems stuck, but Travis is happy, we should merge! =D

amcody commented:
@barentsen Something is going wrong, as we are getting a flat light curve with a particular [non-zero] flux value. But so far we have only made psf photometry attempts on M35, and unfortunately the background prior is hiccuping because it is set to 'nan' in Campaign 0. I guess I will manually specify the bg prior to get around this for now.

amcody commented:
Question: how is the flux prior range being determined? Because the example that I just tried picked a very large range of values (uniform prior). I guess I can dig into the code, but thought you might have a quick answer.

amcody commented:
Okay, with some manual manipulation, we are back to getting something out of prf photometry. It's not pretty compared to my aperture photometry, so we will have to play with detrending to see how much it can be cleaned up.

![index](https://user-images.githubusercontent.com/10542558/42849983-763e5d40-89da-11e8-89be-799460e27be5.png)
